pkg/util/kubernetes/apiserver/resourcetypes.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package apiserver
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/client-go/discovery"
+)
+
+var (
+	resourceCache *ResourceTypeCache
+	cacheOnce     sync.Once
+	cacheErr      error
+)
+
+// ResourceTypeCache is a global cache to store Kubernetes resource types.
+type ResourceTypeCache struct {
+	kindGroupToType map[string]string
+	lock            sync.RWMutex
+	discoveryClient discovery.DiscoveryInterface
+}
+
+// InitializeGlobalResourceTypeCache initializes the global cache if it hasn't been already.
+func InitializeGlobalResourceTypeCache(discoveryClient discovery.DiscoveryInterface) error {
+	cacheOnce.Do(func() {
+		resourceCache = &ResourceTypeCache{
+			kindGroupToType: make(map[string]string),
+			discoveryClient: discoveryClient,
+		}
+
+		// Optionally pre-populate the cache
+		err := resourceCache.prepopulateCache()
+		if err != nil {
+			cacheErr = fmt.Errorf("failed to prepopulate resource type cache: %w", err)
+		}
+	})
+	return cacheErr
+}
+
+// GetResourceType retrieves the resource type for the given kind and group.
+func GetResourceType(kind, apiVersion string) (string, error) {
+	if resourceCache == nil {
+		return "", fmt.Errorf("resource type cache is not initialized")
+	}
+	return resourceCache.getResourceType(kind, apiVersion)
+}
+
+// getResourceType is the instance method to retrieve a resource type.
+func (r *ResourceTypeCache) getResourceType(kind, apiVersion string) (string, error) {
+	group := getAPIGroup(apiVersion)
+	cacheKey := getCacheKey(kind, group)
+
+	// Check the cache
+	r.lock.RLock()
+	resourceType, found := r.kindGroupToType[cacheKey]
+	r.lock.RUnlock()
+	if found {
+		return resourceType, nil
+	}
+
+	// Query the API server and update the cache
+	resourceType, err := r.discoverResourceType(kind, group)
```

Comment:
it can be good to wrap this function with a "backoff retrier" like here:https://github.com/DataDog/datadog-agent/blob/bb8a789b936485812bb302ab41ca471f16ad4106/comp/core/workloadmeta/impl/store.go#L535